native/dtrexc.go

```diff
@@ -0,0 +1,220 @@
+//
+// If compq is lapack.EigDecomp, on return the matrix Q of Schur vectors will be
+// updated by postmultiplying it with Z. If compq is lapack.EigValueOnly, the
+// matrix Q is not referenced and will not be updated. For other values of compq
+// Dtrexc will panic.
+//
+// ifst and ilst specify the reordering of the diagonal blocks of T. The block
+// with row index ifst is moved to row ilst, by a sequence of transpositions
+// between adjacent blocks.
+//
+// If ifst points to the second row of a 2×2 block, ifstOut will point to the
+// first row, otherwise it will be equal to ifst.
+//
+// ilstOut will point to the first row of the block in its final position. If ok
+// is true, ilstOut may differ from ilst by +1 or -1.
+//
+// It must hold that
+//  0 <= ifst < n, and  0 <= ilst < n,
+// otherwise Dtrexc will panic.
+//
+// If ok is false, two adjacent blocks were too close to swap because the
+// problem is very ill-conditioned. T may have been partially reordered, and
+// ilstOut will point to the first row of the block at the position to which it
+// has been moved.
+//
+// work must have length at least n, otherwise Dtrexc will panic.
+//
+// Dtrexc is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dtrexc(compq lapack.EigComp, n int, t []float64, ldt int, q []float64, ldq int, ifst, ilst int, work []float64) (ifstOut, ilstOut int, ok bool) {
+	checkMatrix(n, n, t, ldt)
+	var wantq bool
+	switch compq {
+	default:
+		panic(badEigComp)
+	case lapack.EigValueOnly:
+	case lapack.EigDecomp:
+		wantq = true
+		checkMatrix(n, n, q, ldq)
+	}
+	if (ifst < 0 || n <= ifst) && n > 0 {
+		panic("lapack: ifst out of range")
+	}
+	if (ilst < 0 || n <= ilst) && n > 0 {
+		panic("lapack: ilst out of range")
+	}
+	if len(work) < n {
+		panic(badWork)
+	}
+
+	ok = true
+
+	// Quick return if possible.
+	if n <= 1 {
+		return ifst, ilst, true
+	}
+
+	// Determine the first row of specified block and find out it is 1×1 or 2×2.
+	if ifst > 0 && t[ifst*ldt+ifst-1] != 0 {
+		ifst--
+	}
+	nbf := 1 // Size of the first block.
+	if ifst+1 < n && t[(ifst+1)*ldt+ifst] != 0 {
+		nbf = 2
+	}
+	// Determine the first row of the final block
+	// and find out it is 1×1 or 2×2.
+	if ilst > 0 && t[ilst*ldt+ilst-1] != 0 {
+		ilst--
+	}
+	nbl := 1 // Size of the last block.
+	if ilst+1 < n && t[(ilst+1)*ldt+ilst] != 0 {
+		nbl = 2
+	}
+
+	switch {
+	case ifst == ilst:
+		return ifst, ilst, true
+
+	case ifst < ilst:
+		// Update ilst.
+		if nbf == 2 && nbl == 1 {
+			ilst--
+		}
+		if nbf == 1 && nbl == 2 {
+			ilst++
+		}
+		here := ifst
+		for here < ilst {
+			// Swap block with next one below.
+
+			if nbf == 1 || nbf == 2 {
+				// Current block either 1×1 or 2×2.
+				nbnext := 1 // Size of the next block.
+				if here+nbf+1 < n && t[(here+nbf+1)*ldt+here+nbf] != 0 {
+					nbnext = 2
+				}
+				ok = impl.Dlaexc(wantq, n, t, ldt, q, ldq, here, nbf, nbnext, work)
+				if !ok {
+					return ifst, here, false
```

Comment:
Yeah, I thought about what is the best way when I wrote it and I'm also undecided. Maybe, just maybe, I prefer the current form.